client/src/Tetromino.ts

```diff
@@ -163,34 +171,50 @@ export class Tetromino {
         this.rotation = <0 | 1 | 2 | 3>(rotation % 4);
     }
 
-    // TODO: Properly verify movement is possible before allowing it
-    move(colDelta: number): boolean {
-        const newCol = this.position[1] + colDelta;
-        if (newCol < 0 || newCol >= BOARD_SIZE) {
-            return false;
-        }
-        this.position[1] = newCol;
-        return true;
+    rotateCW() {
+        this.setRotation(this.rotation + 1);
     }
 
-    // TODO: Verify rotation is possible before allowing it
-    private canRotate(): boolean {
-        return true;
+    rotateCCW() {
+        this.setRotation(4 + this.rotation - 1);
     }
 
-    rotateCW(): boolean {
-        if (!this.canRotate()) {
-            return false;
-        }
-        this.setRotation(this.rotation + 1);
-        return true;
-    }
+    moveIfCan(
+        board: Array<Array<TetrominoType | null>>,
+        movement: (tetro: Tetromino) => Tetromino | void
+    ): boolean {
+        let newTetro: Tetromino = cloneDeep(this);
+        const oldTileCoords = newTetro.tiles.map((tile) => [
+            newTetro.position[0] + tile[0],
+            newTetro.position[1] + tile[1],
+        ]);
 
-    rotateCCW(): boolean {
-        if (!this.canRotate()) {
-            return false;
+        // look-ahead for the next tetromino state after movement
+        newTetro = movement(newTetro) || newTetro;
+        for (let i = 0; i < this.tiles.length; i++) {
+            const [row, col] = newTetro.tiles[i];
+
+            if (
+                // 1. the new position has some other tiles in it
+                board[newTetro.position[0] + row][newTetro.position[1] + col] !=
+                    null &&
+                // AND 2. the tiles are NOT from the old self
+                // TODO falling out of bounds
+                // TODO hitting outside the visible arena (game over)
+                !oldTileCoords.some(
+                    ([oldRow, oldCol]) =>
+                        newTetro.position[0] + row == oldRow &&
+                        newTetro.position[1] + col == oldCol
+                )
```

Comment:
I think these should be computed as boolean variables before the if just for readability